### PR TITLE
Refactor Menus

### PR DESCRIPTION
**Pull request by Lord-Kamina**
Our current code is essentially creating new Menu and MenuOption objects permanently, when there's no real good reason to do that.

This was originally part of a humoungous PR involving overhauling the text rendering, but I am in the process of splitting it in usable PRs.

### More

- [ ] Added/updated documentation

### CODE REVIEW

**Comment by ooshlablu** [comment]
Tested with singing and guitar/bass on a few songs. I didn't notice anything different game play or experience wise. :+1:

File: game/dancegraph.cc [hunk 1]
```diff
@@ -103,7 +103,8 @@ void DanceGraph::setupJoinMenu() {
 	m_menu.clear();
 	updateJoinMenu();
 	// Populate root menu
-	m_menu.add(MenuOption(_("Ready!"), _("Start performing!")));
+	auto _ready = std::make_unique<MenuOption>(_("Ready!"), _("Start performing!"));
```
Comment (Baklap4) [inline game/dancegraph.cc:106]: please get rid of the `_` prefix for local variables.

File: game/screen_songs.cc [hunk 2]
```diff
@@ -529,17 +529,16 @@ void ScreenSongs::drawMenu() {
 	for (MenuOptions::const_iterator it = m_menu.begin(); it != m_menu.end(); ++it) {
```
Comment (OznOg) [inline game/screen_songs.cc:529]: use range for loop to get rid of (*it)

File: game/menu.hh [hunk 3]
```diff
@@ -95,18 +95,17 @@ public:
...
 	const MenuOptions::const_iterator begin() const { return menu_stack.back()->begin(); }
 	const MenuOptions::const_iterator end() const { return menu_stack.back()->end(); }
-	const MenuOptions getOptions() const { return *menu_stack.back(); }
+	const MenuOptions& getOptions() const { return *menu_stack.back(); }
 
 	Dimensions dimensions;
 
 private:
-	MenuOptions root_options;
+	bool m_open;
```
Comment (OznOg) [inline game/menu.hh:107]: why did  m_open moved here?

File: game/menu.cc [hunk 4]
```diff
@@ -4,9 +4,7 @@
 #include "fs.hh"
 #include "game.hh"
 
-MenuOption::MenuOption(std::string const& nm, std::string const& comm, MenuImage img):
-  type(), value(), newValue(), callback(), image(img), name(nm), comment(comm), namePtr(), commentPtr()
-{}
+MenuOption::MenuOption(std::string const& nm, std::string const& comm, MenuImage img): type(), value(), newValue(), options(std::make_shared<MenuOptions>()), callback(), image(img), name(nm), comment(comm), namePtr(), commentPtr() {}
```
Comment (OznOg) [inline game/menu.cc:7]: no need to initialize a shared_ptr to null, it is already done in the defautl constructor

File: game/menu.cc [hunk 5]
```diff
@@ -31,10 +29,10 @@ bool MenuOption::isActive() const {
 }
 
 
-Menu::Menu(): dimensions(), m_open(true) { clear(); }
+Menu::Menu(): dimensions(), m_open(true), root_options(std::make_shared<MenuOptions>()) { clear(); }
```
Comment (OznOg) [inline game/menu.cc:32]: same, If yo ureally want to stress this is set to nullptr, set it at nullptr at its declaration.

File: game/menu.cc [hunk 6]
```diff
@@ -92,10 +90,10 @@ void Menu::action(int dir) {
 }
 
 void Menu::clear(bool save_root) {
-	if (!save_root) root_options.clear();
+	if (!save_root) root_options->clear();
 	menu_stack.clear();
 	selection_stack.clear();
-	menu_stack.push_back(&root_options);
```
Comment (OznOg) [inline game/menu.cc:98]: not sure this is realy what we want to have: should menu_stack **own** the root_options? It seems to me the sub menu is just using it, so using a raw pointer can be ok (or reference wrapper if you really don't like pointers), or if you really fear dangling pointers, you may use weak_ptr here.

File: game/menu.hh [hunk 7]
```diff
@@ -37,7 +37,7 @@ public:
 	/// Make the option set a given value for ConfigItem and close the menu.
 	MenuOption& setter(ConfigItem& val, ConfigItem newval) { type = Type::SET_AND_CLOSE; value = &val; newValue = newval; return *this; }
 	/// Make the option open a submenu
-	MenuOption& submenu(MenuOptions opts) { type = Type::OPEN_SUBMENU; options = opts; return *this; }
+	MenuOption& submenu(MenuOptions opts) { type = Type::OPEN_SUBMENU; options = std::make_shared<MenuOptions>(std::move(opts)); return *this; }
```
Comment (OznOg) [inline game/menu.hh:40]: option =  std::make_shared<MenuOptions>(std::move(opts)) => option = std::move(opts);

File: game/menu.hh [hunk 8]
```diff
@@ -95,18 +95,17 @@ public:
...
 	const MenuOptions::const_iterator end() const { return menu_stack.back()->end(); }
-	const MenuOptions getOptions() const { return *menu_stack.back(); }
+	const MenuOptions& getOptions() const { return *menu_stack.back(); }
 
 	Dimensions dimensions;
 
 private:
-	MenuOptions root_options;
+	bool m_open;
+	std::shared_ptr<MenuOptions> root_options;
```
Comment (OznOg) [inline game/menu.hh:108]: std::shared_ptr<MenuOptions> root_options = nullptr; If you really want to explicitely show it was initialized

File: game/screen_songs.cc [hunk 9]
```diff
@@ -514,7 +514,7 @@ void ScreenSongs::drawMenu() {
 	if (m_menu.empty()) return;
 	// Some helper vars
 	ThemeInstrumentMenu& th = *m_menuTheme;
-	const auto cur = &m_menu.current();
+	MenuOption& cur = m_menu.current();
```
Comment (OznOg) [inline game/screen_songs.cc:517]: why removing the const and the auto?

File: game/screen_songs.cc [hunk 10]
```diff
@@ -529,17 +529,16 @@ void ScreenSongs::drawMenu() {
 	for (MenuOptions::const_iterator it = m_menu.begin(); it != m_menu.end(); ++it) {
 		// Pick the font object
 		SvgTxtTheme* txt = &th.option_selected;
-		if (cur != &*it)
-			txt = &(th.getCachedOption(it->getName()));
+		if (&cur != &*it->get()) txt = &(th.getCachedOption((*it)->getName()));
```
Comment (OznOg) [inline game/screen_songs.cc:532]: if (&cur != it->get())

File: game/instrumentgraph.cc [hunk 11]
```diff
@@ -53,10 +53,17 @@ bool InstrumentGraph::dead() const { return m_jointime != m_jointime || m_dead >
 
 void InstrumentGraph::setupPauseMenu() {
 	m_menu.clear();
-	m_menu.add(MenuOption(_("Resume"), _("Back to performing!")));
-	m_menu.add(MenuOption(_("Rejoin"), _("Change selections")).changer(m_rejoin));
-	m_menu.add(MenuOption(_("Restart"), _("Start the song\nfrom the beginning")).screen("Sing"));
-	m_menu.add(MenuOption(_("Quit"), _("Exit to song browser")).screen("Songs"));
+	auto _resume = std::make_unique<MenuOption>(_("Resume"), _("Back to performing!"));
```
Comment (OznOg) [inline game/instrumentgraph.cc]: if you don't use _resume, add call directly make_unique in the m_menu.add

File: game/guitargraph.cc [hunk 12]
```diff
@@ -160,13 +160,17 @@ void GuitarGraph::setupJoinMenuDifficulty() {
 	}
 	m_selectedDifficulty = ConfigItem(ol); // Create a ConfigItem from the option list
 	m_selectedDifficulty.select(cur); // Set the selection to current level
-	m_menu.add(MenuOption("", _("Select difficulty")).changer(m_selectedDifficulty)); // MenuOption that cycles the options
+	auto _selectDifficulty = std::make_unique<MenuOption>("", _("Select difficulty"));
```
Comment (OznOg) [inline game/guitargraph.cc]: m_menu.add(std::make_unique<MenuOption>("", _("Select difficulty")).changer(m_selectedDifficulty));  ? why do you need to split that in 3 lines ?

File: game/guitargraph.cc [hunk 13]
```diff
@@ -160,13 +160,17 @@ void GuitarGraph::setupJoinMenuDifficulty() {
...
+	_selectDifficulty->changer(m_selectedDifficulty);
+	m_menu.add(std::move(_selectDifficulty)); // MenuOption that cycles the options
 	m_menu.back().setDynamicName(m_difficultyOpt); // Set the title to be dynamic
 }
 
 void GuitarGraph::setupJoinMenuDrums() {
 	setupJoinMenuDifficulty();
-	m_menu.add(MenuOption(_("Lefty-mode"), "").changer(m_leftymode));
+	auto _leftyMode = std::make_unique<MenuOption>(_("Lefty-mode"), "");
+	_leftyMode->changer(m_leftymode);
```
Comment (OznOg) [inline game/guitargraph.cc]: same

**Comment by OznOg** [comment]
I don't really get the purpose of this patch. You seem to talk about not recreating things, but looks like it tries to fix some ownership problem on menu options. Could you give some more context please?
BTW, I think that using unique_ptr in menu intefaces is making the code harder to read, if you really want to use shared ptr in internal, maybe you should use some pimpl idom to "hide" it from user.

**Comment by Lord-Kamina** [comment]
> I don't really get the purpose of this patch. You seem to talk about not recreating things, but looks like it tries to fix some ownership problem on menu options. Could you give some more context please? BTW, I think that using unique_ptr in menu intefaces is making the code harder to read, if you really want to use shared ptr in internal, maybe you should use some pimpl idom to "hide" it from user.

As discussed on Discord, options are being copied (which is what I was missing in my explanation) a bazillion times on master. The PR makes it so each option is built once and destroyed once, when entering and quitting the screen. 

You said this is chiefly because of my usage std::move, and that we could scrap the smart pointers. Sounds valid, I can't entirely remember whether the pointers themselves were relevant for the rest of the PR where I originally did this.

I'm currently trying to divide that into useable chunks

**Pull request closed without merging**